### PR TITLE
Pzucker/inline sub

### PR DESCRIPTION
**Pull request by philzook58**
Consolidated mk_default_env and mk_inline_env into a single function.
In addition, any function names starting with "subs_" should be inclined by default

### CODE REVIEW

File: wp/lib/bap_wp/tests/unit/test_compare.ml [hunk 1]
```diff
@@ -68,8 +68,8 @@ let test_block_pair_1 (test_ctx : test_ctxt) : unit =
 let test_block_pair_2 (test_ctx : test_ctxt) : unit =
   let ctx = Env.mk_ctx () in
   let var_gen = Env.mk_var_gen () in
-  let env1 = Pre.mk_default_env ctx var_gen in
-  let env2 = Pre.mk_default_env ctx var_gen in
+  let env1 = Pre.mk_env ctx var_gen  in
```
Comment (fortunac) [inline wp/lib/bap_wp/tests/unit/test_compare.ml]: Any reason for the extra space at the end?

File: wp/plugin/wp.ml [hunk 2]
```diff
@@ -76,18 +81,21 @@ let analyze_proj (proj : project) (var_gen : Env.var_gen) (ctx : Z3.context)
...
-        if List.is_empty to_inline then
-          subs
-        else
-          to_inline
-          |> List.map ~f:(find_func_err subs)
-          |> Seq.of_list
+      let to_inline' = if (inline || has_inline_funcs to_inline) then
+        begin
+          if List.is_empty to_inline then
+            (Printf.printf "using all subs"; subs)
```
Comment (fortunac) [inline wp/plugin/wp.ml]: Maybe make this in `info` log instead of a print?

File: wp/plugin/wp.ml [hunk 3]
```diff
@@ -76,18 +81,21 @@ let analyze_proj (proj : project) (var_gen : Env.var_gen) (ctx : Z3.context)
...
+            |> List.map ~f:(find_func_err subs)
+            |> Seq.of_list
+            end
+        end
+        else Seq.empty
       in
-      Pre.mk_inline_env ctx var_gen ~subs ~arch ~to_inline:to_inline
-    else
-      Pre.mk_default_env ctx var_gen ~subs ~arch
+      Seq.iter (filter_subs subs) ~f:(fun sub -> Printf.printf "%s\n" (Sub.name sub));
```
Comment (fortunac) [inline wp/plugin/wp.ml]: What is this supposed to be printing out?

**Comment by philzook58** [comment]
These are good catches

File: wp/plugin/wp.ml [hunk 4]
```diff
@@ -134,14 +106,10 @@ let compare_projs (proj : project) (file1: string) (file2 : string)
...
-                      find_inline_funcs to_inline ~to_find:subs1 ~to_check:subs2
-                      else Seq.empty in
-    let to_inline2 = if inline || has_inline_funcs to_inline then 
-                      find_inline_funcs to_inline ~to_find:subs2 ~to_check:subs1
-                      else Seq.empty in
-    Pre.mk_env ctx var_gen ~subs:subs1 ~arch:arch ~to_inline:(Seq.append to_inline1 (filter_subs subs1)),
-    Pre.mk_env ctx var_gen ~subs:subs2 ~arch:arch ~to_inline:(Seq.append to_inline2 (filter_subs subs2))
+    let to_inline1 = match_inline to_inline subs1 in 
+    let to_inline2 = match_inline to_inline subs2 in 
+    Pre.mk_env ctx var_gen ~subs:subs1 ~arch:arch ~to_inline:to_inline1,  (* (Seq.append to_inline1 (filter_subs subs1)) *)
```
Comment (fortunac) [inline wp/plugin/wp.ml]: Can you remove the commented out code?

**Comment by philzook58** [comment]
Changed to use regular expressions to specify functions to inline
Does add a dependency for the Re library
I changed the inline flag to be the one that takes the regular expression
the old bool taking flag no longer exists
documentation is updated accordingly
and --wp-inline-funcs no longer exists 
--wp-inline=.* will inline everything

File: wp/plugin/README.md [hunk 5]
```diff
@@ -9,6 +9,7 @@ Depends on:
     - bap 2.0.0-alpha
     - ounit 2.0.8
     - z3 4.8.6
+    - re
```
Comment (fortunac) [inline wp/plugin/README.md]: Is there a specific version for `re` we're using?

File: wp/plugin/tests/test_wp.ml [hunk 6]
```diff
@@ -107,30 +103,27 @@ let suite = [
   "Switch Case Assignments"        >:: test_compare_elf "switch_case_assignments" "SAT!" ~func:"process_status";
   "Switch Cases"                   >:: test_compare_elf "switch_cases" "SAT!" ~func:"process_message" ~check_calls:true;
   "No Stack Protection"            >:: test_compare_elf "no_stack_protection" "SAT!" ~output_vars:"RSI,RAX";
-  "Retrowrite Stub"                >:: test_compare_elf "retrowrite_stub" "UNSAT!"
-    ~inline:true ~to_inline:"__afl_maybe_log";
-  "Retrowrite Stub: imply inline"  >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~to_inline:"__afl_maybe_log";
-  "Retrowrite Stub: inline all"    >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~inline:true;
+  "Retrowrite Stub"                >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~inline:"__afl_maybe_log";
+  "Retrowrite Stub: imply inline"  >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~inline:"__afl_maybe_log";
```
Comment (fortunac) [inline wp/plugin/tests/test_wp.ml]: I think we can remove the "imply inline" tests, now that we consolidated the inline flag.

File: wp/plugin/tests/test_wp.ml [hunk 7]
```diff
@@ -107,30 +103,27 @@ let suite = [
...
-    ~inline:true ~to_inline:"foo";
-  "Function Call: imply inline"    >:: test_single_elf "function_call" "main" "SAT!" ~to_inline:"foo";
-  "Function Call: inline all"      >:: test_single_elf "function_call" "main" "SAT!" ~inline:true;
-  "Function Spec"                  >:: test_single_elf "function_spec" "main" "UNSAT!"
-    ~inline:true ~to_inline:"foo";
-  "Function Spec: imply inline"    >:: test_single_elf "function_spec" "main" "UNSAT!" ~to_inline:"foo";
-  "Function Spec: inline all "     >:: test_single_elf "function_spec" "main" "UNSAT!" ~inline:true;
-  "Function Spec: no inlining"     >:: test_single_elf "function_spec" "main" "SAT!" ~inline:false;
+  "Function Call"                  >:: test_single_elf "function_call" "main" "SAT!" ~inline:"foo";
+  "Function Call: imply inline"    >:: test_single_elf "function_call" "main" "SAT!" ~inline:"foo";
```
Comment (fortunac) [inline wp/plugin/tests/test_wp.ml]: Here too.

File: wp/plugin/tests/test_wp.ml [hunk 8]
```diff
@@ -107,30 +103,27 @@ let suite = [
...
+  "Function Call: inline all"      >:: test_single_elf "function_call" "main" "SAT!" ~inline:".*";
+  "Function Spec"                  >:: test_single_elf "function_spec" "main" "UNSAT!" ~inline:"foo";
+  "Function Spec: imply inline"    >:: test_single_elf "function_spec" "main" "UNSAT!" ~inline:"foo";
```
Comment (fortunac) [inline wp/plugin/tests/test_wp.ml]: Here too.

File: wp/plugin/wp.ml [hunk 9]
```diff
@@ -126,14 +106,10 @@ let compare_projs (proj : project) (file1: string) (file2 : string)
...
-    if inline || has_inline_funcs to_inline then
-      let to_inline1 = find_inline_funcs to_inline ~to_find:subs1 ~to_check:subs2 in
-      let to_inline2 = find_inline_funcs to_inline ~to_find:subs2 ~to_check:subs1 in
-      Pre.mk_inline_env ctx var_gen ~subs:subs1 ~arch:arch ~to_inline:to_inline1,
-      Pre.mk_inline_env ctx var_gen ~subs:subs2 ~arch:arch ~to_inline:to_inline2
-    else
-      Pre.mk_default_env ctx var_gen ~subs:subs1 ~arch:arch,
-      Pre.mk_default_env ctx var_gen ~subs:subs2 ~arch:arch
+    let to_inline1 = match_inline to_inline subs1 in 
+    let to_inline2 = match_inline to_inline subs2 in
```
Comment (fortunac) [inline wp/plugin/wp.ml]: Do you think it's a good idea to have an info log about which functions are going to be inlined?

File: wp/plugin/wp.ml [hunk 10]
```diff
@@ -47,48 +47,31 @@ let update_default_num_unroll (num_unroll : int option) : unit =
...
-    to_find
-  else
-    inline_funcs
-    |> List.map ~f:(find_func_in_one_of ~to_find ~to_check)
-    |> List.concat
-    |> Seq.of_list
+let match_inline (to_inline : string option) (subs : (Sub.t Seq.t)) : Sub.t Seq.t =
+  match to_inline with
+  | None -> Seq.empty
+  | Some to_inline -> let inline_pat = Re.Posix.re to_inline |> Re.Posix.compile in
```
Comment (fortunac) [inline wp/plugin/wp.ml]: Do you think it's a good idea to have a warning log in the case we couldn't match on anything?

**Comment by fortunac** [comment]
Tests are passing for me, and everything looks super good! :tada:

File: wp/plugin/wp.ml (hunk 10 above)
Comment (philzook58) [reply wp/plugin/wp.ml]: Smart.

File: wp/plugin/wp.ml (hunk 9 above)
Comment (philzook58) [reply wp/plugin/wp.ml]: Smart.

File: wp/plugin/tests/test_wp.ml (hunk 8 above)
Comment (philzook58) [reply wp/plugin/tests/test_wp.ml]: Good catch

File: wp/plugin/README.md (hunk 5 above)
Comment (philzook58) [reply wp/plugin/README.md]: Added 1.9.0

File: wp/lib/bap_wp/tests/unit/test_compare.ml [hunk 11]
```diff
@@ -391,9 +391,9 @@ let test_sub_pair_fun_3 (test_ctx : test_ctxt) : unit =
   let blk3 = blk3 |> mk_call (Label.direct (Term.tid blk4)) (Label.direct (Term.tid call2_sub)) in
   let main_sub1 = mk_sub ~tid:sub1_tid ~name:"main_sub" [blk1; blk2; blk3; blk4] in
   let main_sub2 = mk_sub ~tid:sub2_tid ~name:"main_sub" [blk1'; blk2; blk3; blk4] in
-  let env1 = Pre.mk_default_env ctx var_gen
+  let env1 = Pre.mk_env ctx var_gen
```
Comment (codyroux) [inline wp/lib/bap_wp/tests/unit/test_compare.ml]: Can you install `ocp-indent` pretty please?

File: wp/plugin/wp.ml (hunk 10 above)
Comment (codyroux) [reply wp/plugin/wp.ml]: I do!

**Comment by codyroux** [comment]
This looks really great! Good job Phil! Sorry to bug you with more minor comments.

**Pull request merged**